### PR TITLE
Harmony 1238 stuck jobs bug

### PR DESCRIPTION
**Pull request by indiejames**
## Jira Issue ID
HARMONY-1238

## Description
Makes updates to work-items asynchronous to prevent services having to wait. This allows locking the jobs table to avoid race conditions on updates since the work-item update can take longer without bogging down services.

## Local Test Steps
Can't force the race condition this is supposed to eliminate - best attempt is to run a few really long queries and make sure they don't get in a state where all the work items are done, but the job is not.

## PR Acceptance Checklist
* [x] Acceptance criteria met
* [x] Tests added/updated (if needed) and passing
* [ ] ~Documentation updated (if needed)~

### CODE REVIEW

File: app/backends/workflow-orchestration.ts [hunk 1]
```diff
@@ -590,8 +584,30 @@ export async function updateWorkItem(req: HarmonyRequest, res: Response): Promis
...
+export async function updateWorkItem(req: HarmonyRequest, res: Response): Promise<void> {
+  const { id } = req.params;
+  const { status, hits, results, scrollID, errorMessage } = req.body;
+  const totalGranulesSize = req.body.totalGranulesSize ? parseFloat(req.body.totalGranulesSize) : 0;
+
+  const update = new WorkItemUpdate(
+    { workItemID: id, status, hits, serializedResults: JSON.stringify(results), scrollID, errorMessage, totalGranulesSize },
+  );
+
+  // asynchronously handle the update
```
Comment (chris-durbin) [inline app/backends/workflow-orchestration.ts:607]: Give us more info on the comment on why we are doing this - what problem are we looking to address? What happens if the asynchronous update fails, and why is it ok that we ignore it?

File: app/models/work-item-interface.ts [hunk 2]
```diff
@@ -63,6 +63,35 @@ export interface WorkItemRecord {
   createdAt: Date;
 }
 
+/**
+ * Interface the defines the fields in a work-item update from a service
```
Comment (chris-durbin) [inline app/models/work-item-interface.ts:67]: Interface *that* defines

File: app/models/work-item-interface.ts [hunk 3]
```diff
@@ -63,6 +63,35 @@ export interface WorkItemRecord {
   createdAt: Date;
 }
 
+/**
```
Comment (chris-durbin) [inline app/models/work-item-interface.ts:66]: I just realized - I don't think this interface or the database record are actually used (no SQL updates or migrations in here). I would just leave those files in a branch for you somewhere if you think we might want to track the updates in a database table at some point. Pushing them to main will make people think this code is being used.

**Review by chris-durbin — COMMENTED** [review]
I haven't tested yet.

File: app/models/work-item-interface.ts (hunk 3 above)
Comment (indiejames) [reply app/models/work-item-interface.ts:66]: I thought you said to leave them in when we talked about it the other day. I'll pull them out.

File: app/backends/workflow-orchestration.ts (hunk 1 above)
Comment (indiejames) [reply app/backends/workflow-orchestration.ts:607]: Done.

File: test/work-items/work-backends.ts [hunk 4]
```diff
@@ -281,13 +281,12 @@ describe('Work Backends', function () {
...
-              this.job.status =
-                this.workItem.status = updateState;
-              this.response = await updateWorkItem(this.backend, this.workItem);
+              this.workItem.status = updateState;
+              await updateWorkItem(this.backend, this.workItem);
             });
-            it('fails the update', function () {
-              expect(this.response.status).to.equal(409);
-              expect(this.response.text).to.equal(`Job was already ${terminalState}.`);
+            it('fails the update', async function () {
```
Comment (vinnyinverso) [inline test/work-items/work-backends.ts:287]: Not sure if "fails the update" accurately describes what's being tested here anymore?

Comment (vinnyinverso) [reply test/work-items/work-backends.ts:287]: Never mind, I think I get the gist!

Comment (indiejames) [reply test/work-items/work-backends.ts:287]: I just re-worded it to "fails to update the work item"

File: app/models/work-item-interface.ts (hunk 3 above)
Comment (indiejames) [reply app/models/work-item-interface.ts:66]: Done.

File: app/models/work-item-interface.ts (hunk 2 above)
Comment (indiejames) [reply app/models/work-item-interface.ts:67]: Removed.

**Review by vinnyinverso — APPROVED** [review]

**Review by chris-durbin — APPROVED** [review]
Tested in a sandbox environment. Database performance looked good - there were a minimal number of slow queries logged and all under two seconds. The CPU load on the database was extremely low and only required a database capacity of 2 units.

**Pull request merged**